src/Push.cpp

```diff
@@ -0,0 +1,199 @@
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_SNDTIMEO,
+                 &timeoutInMs,
+                 sizeof(timeoutInMs));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect) {
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_connect");
+   } else {
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_bind");
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+             << " socket at: " << mProtocolHandler->GetLocation()
+             << " with timeout: " << timeoutInMs;
+}
+/**
+ * Construct a specific push queue
+ */
+Push::Push(const std::string& location,
+           const int timeoutInMs,
+           const int bufferSize,
+           const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PUSH);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   if (timeoutInMs > 0) {
+      nn_setsockopt(mSocket,
+                    NN_SOL_SOCKET,
+                    NN_SNDTIMEO,
+                    &timeoutInMs,
+                    sizeof(timeoutInMs));
+   } else {
+      LOG(INFO) << " zero or negative timeout requested, blocking socket";
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_SNDBUF,
+                 &bufferSize,
+                 sizeof(bufferSize));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect) {
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_connect");
+   } else {
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = std::string("nn_bind");
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
```

Comment:
In our C++ guideline we discuss throwing exceptions at length, http://github.schq.secious.com/Logrhythm/CppGuide/blob/master/cppguide.xml 

Please read through it. The common pattern we use in our code is to **not** throw in constructors but instead using the factory pattern. 

The downside is that that requires heap allocation. Here (Research: http://github.schq.secious.com/Logrhythm/Research/pull/1609/files ) we use heap allocation anyways so that downside can be ignored.